### PR TITLE
Go unit tests: Determine coverage for outside packages as well.

### PR DESCRIPTION
**Pull request by michael-berlin**
@enisoc @alainjobart 

**TL;DR: 6% higher coverage at the expense of 4 minute extra test time.**

By default, go test -cover only instruments code within the same package. Therefore, code which is only tested by integration tests (located in different/outside packages) isn't counted at all.

To solve this problem, the go test -coverpkg flag can be used to increase the scope of instrumentation. I've changed the goveralls script such that for each package under test, all dependent packages will be instrumented.

For each package, this results in a cover profile. All cover profiles have to be merged and then uploaded to coveralls.io.

There are two ways to merge the profiles:
a) use gocov which uses its own JSON format as output
b) use gocovmerge which produces a cover profile that can be used by other Go tools (e.g. the HTML visualization)

## Summary of Results
(each merge approach considered separately):

### Example runs on coveralls.io:
old per-package coverage: https://coveralls.io/builds/3459978
full coverage, gocov JSON merge: https://coveralls.io/builds/3461277
full coverage, gocovmerge: https://coveralls.io/builds/3461299

### total coverage:
```
old per-package: ~56%
JSON merge:      ~82%
gocovmerge:      ~62%
```

### relevant lines:
```
old per-package: ~47k
JSON merge:      ~71k
gocovmerge:      ~62k
```

### runtime increase:
```
old per-package: ~2m42s
new dep-package: ~6m48s
```

The downside of merge approach a) is that it drastically changes the way coverage is counted. For example, all whitespace is now counted as relevant lines and therefore the coverage increases with the ratio between whitespace/total lines ;-) Therefore, I suggest to go with approach b).

Merge approach b) also increases the number of relevant lines because: The old coverage only counted packages with *_test.go files. Now, also packages without *_test.go files are counted if they are dependency of a package with tests. Examples are: go/vt/schema/schema.go or go/vt/topo/test/vschema.go

Also note that we cannot instrument all our code (which would give us the maximum number of relevant lines) because for example the go/cmd/* packages have flags with the same name defined as global variables and that let's make the compiler stutter. In addition to that, it would be much slower because the result of the instrumentation doesn't get cached.

If we plan to go forward with this pull request, we'll need to rebalance the test shards on Travis.

### CODE REVIEW

**Comment by enisoc** [comment]
Thanks for the detailed analysis. So it seems to come down to a cost of 4min extra runtime. We could probably absorb that now, but we may start wanting that 4min back later when we add mysql56 and etcd to Travis. What do you think?

**Comment by alainjobart** [comment]
Thanks for doing this, but I'm not sure it's worth the extra effort, especially since inside google3 the coverage tool does the right thing?

**Comment by michael-berlin** [comment]
Closing this PR as we decided to not move forward with it. It's too time consuming and we have more accurate internal coverage numbers already.

Nonetheless, I'll later refactor this code to run coverage on a specific package and get the coverage for all packages. This will help us for integration tests which cover code outside the same package.
